[Enhancement]Support batch drop partitions syntax

leiyang0324 opened the pull request:
## Why I'm doing:
The current latest version lacks the functionality for batch deletion of partitions. When a large number of partitions need to be deleted, the efficiency is relatively low due to the necessity of obtaining locks during deletion.

## What I'm doing:
Add batch deletion syntax and functionality for formal partitions and temporary partitions."

```
-- formal partition
ALTER TABLE test.test_broke_load_index_vip DROP PARTITIONS IF EXISTS START("2024-02-24") END("2024-02-26") EVERY (INTERVAL 1 DAY);
-- temporary partition
ALTER TABLE test.test_broke_load_index_vip DROP TEMPORARY PARTITIONS IF EXISTS START("2024-01-01") END("2024-01-05") EVERY (INTERVAL 1 DAY);
```

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

leiyang0324 committed:
[Enhancement]Support batch drop partitions syntax

leiyang0324 committed:
[Enhancement]Support batch drop partitions syntax

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/StarRocks/starrocks?pullRequest=41739) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/StarRocks/starrocks?pullRequest=41739) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/StarRocks/starrocks?pullRequest=41739) it.</sub>

leiyang0324 committed:
[fix cr]Support batch drop partitions syntax

leiyang0324 committed:
[fix cr]Support batch drop partitions syntax

leiyang0324 committed:
[code style format]Support batch drop partitions syntax

leiyang0324 committed:
[code style format]Support batch drop partitions syntax

nshangyiming commented:
@leiyang0324 Excellent job, @gengjun-git and i will help to review this pr, thanks.

gengjun-git commented:
@leiyang0324 Can you give an example of the new syntax?

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1589,59 +1589,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+            if (partitionDesc instanceof MultiRangePartitionDesc) {
+
+                if (!(partitionInfo instanceof RangePartitionInfo)) {
+                    throw new DdlException("Batch creation of partitions only support range partition tables.");
```
`throw new DdlException("Batch creation of partitions only support range partition tables.");`
should be 
`throw new DdlException("Batch deletion of partitions only support range partition tables.");`

leiyang0324 replied:
yes, I will fix this issue

leiyang0324 commented:
@gengjun-git

leiyang0324 commented:
@gengjun-git  Thank you for your reply
-- formal partition
ALTER TABLE test.test_broke_load_index_vip  DROP PARTITIONS IF EXISTS START("2024-02-24") END("2024-02-26") EVERY (INTERVAL 1 DAY);
-- temporary partition
ALTER TABLE test.test_broke_load_index_vip  DROP TEMPORARY PARTITIONS IF EXISTS START("2024-01-01") END("2024-01-05") EVERY (INTERVAL 1 DAY);

leiyang0324 committed:
Modify Exception Message Content

leiyang0324 committed:
Modify Exception Message Content

nshangyiming commented on fe/fe-core/src/test/java/com/starrocks/sql/analyzer/AstBuilderTest.java:
```diff
@@ -92,4 +94,66 @@ public void testTruncatePartition() throws Exception {
         Assert.assertTrue(c.getPartitionNames().getPartitionNames().get(0).equals("p1"));
     }
+
+    @Test
```
Except UT for analyzer, plz add more UT for execution, you can refer to `fe/fe-core/src/test/java/com/starrocks/catalog/DropPartitionTest.java`.

nshangyiming replied:
We have a 80% coverage requirement.

leiyang0324 replied:
Alright, I will add unit tests.

gengjun-git commented:
> @gengjun-git Thank you for your reply -- formal partition ALTER TABLE test.test_broke_load_index_vip DROP PARTITIONS IF EXISTS START("2024-02-24") END("2024-02-26") EVERY (INTERVAL 1 DAY); -- temporary partition ALTER TABLE test.test_broke_load_index_vip DROP TEMPORARY PARTITIONS IF EXISTS START("2024-01-01") END("2024-01-05") EVERY (INTERVAL 1 DAY);

@wangsimo0 @jaogoy PTAL, do you think this syntax is reasonable?

nshangyiming commented:
@wangsimo0 plz give some advice about the syntax

gengjun-git commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1655,13 +1716,10 @@ public void dropPartition(Database db, Table table, DropPartitionClause clause)
-                clause.isForceDrop());
+        DropPartitionInfo info = new DropPartitionInfo(db.getId(), olapTable.getId(),
+                isTempPartition, clause.isForceDrop(), partitionNames);
         GlobalStateMgr.getCurrentState().getEditLog().logDropPartition(info);
```
No, you are correct here, but you should handle the batch partitions when replaying DropPartitionInfo at `LocalMetastore.replayDropPartition`

leiyang0324 replied:
> No, you are correct here, but you should handle the batch partitions when replaying DropPartitionInfo at `LocalMetastore.replayDropPartition`

If I don't need to process each partition name to editlog individually within the com.starrocks.server.LocalMetastore#dropPartition method, then I would need to handle this structure during replay in LocalMetastore.replayDropPartition. However, if I use a loop to handle editlog in com.starrocks.server.LocalMetastore#dropPartition, then I don't need to change the content of the LocalMetastore.replayDropPartition method.

nshangyiming commented:
@mergify rebase

Smith-Cruise committed:
[Enhancement] Merge multiply blocks into one io request in CacheInputStream (#38882)

Signed-off-by: Smith Cruise <chendingchao1@126.com>

Commit:
[Enhancement] Add COREDUMP_COLLECT_MINIMAL_INTERVAL to throttle the core dump collection of too frequent crash (#43013)

Signed-off-by: Deng Liu <d.liu@celonis.com>

ABingHuang committed:
[BugFix] fix partition prune for list partition on ne expression (#42913)

Signed-off-by: ABingHuang <codekhuang@163.com>

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+
+                List<Column> partitionColumns = rangePartitionInfo.getPartitionColumns();
+                if (partitionColumns.size() != 1) {
+                    throw new DdlException("Alter batch drop partition only support single range column.");
```
It's better to maintain unity of style as above, and print more useful info, like:
`Batch deletion of partitions only support range partition tables with only a column, current column num is  {$partitionColumns.size()}: ${partitionColumns()}.`

Lrabbit125 committed:
[Feature] BE support IPv6 (#41554)

Signed-off-by: lrabbit <2149189364@qq.com>
Signed-off-by: Astralidea <astralidea@163.com>
Signed-off-by: Xueyan Li <astralidea@163.com>
Co-authored-by: linrui <linrui20@huawei.com>
Co-authored-by: Astralidea <astralidea@163.com>

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+            if (partitionDesc instanceof MultiRangePartitionDesc) {
+
+                if (!(partitionInfo instanceof RangePartitionInfo)) {
+                    throw new DdlException("Batch deletion of partitions only support range partition tables.");
```
It's recommended to put those error messages shown to users into ErrorCode.java.
and the SQLState could be 42000.

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+                MultiRangePartitionDesc multiRangePartitionDesc = (MultiRangePartitionDesc) partitionDesc;
+                boolean isAutoPartitionTable = false;
+                if (partitionInfo.getType() == PartitionType.EXPR_RANGE) {
+                    isAutoPartitionTable = true;
```
If it's clear for every kind of partition type, `isAutoPartition` should be a {const} field/method of the PartitionInfo.
cc @nshangyiming

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+                context.setProperties(properties);
+                context.setTempPartition(clause.isTempPartition());
+                List<SingleRangePartitionDesc> singleRangePartitionDescs = multiRangePartitionDesc.convertToSingle(context);
+                partitionNames.addAll(singleRangePartitionDescs.stream()
```
Shell we take consideration of the number of partitions? If it's large, will it be OK?

leiyang0324 replied:
The time complexity of this method is O(N), and I believe there are no performance issues within a reasonable time frame.

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+        } else {
+            partitionNames.add(clause.getPartitionName());
+        }
+        List<String> existPartition = Lists.newArrayList();
```
It would be better to add `s` for plural form.

leiyang0324 replied:
ok

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+        List<String> notExistPartition = Lists.newArrayList();
+
+        for (String partitionName : partitionNames) {
+            if (olapTable.checkPartitionNameExist(partitionName, isTempPartition)) {
```
Shell we task consideration of atomicity of fetch all partitions?

leiyang0324 replied:
"I have reviewed the logic for bulk partition addition and found that it currently lacks atomicity. If atomicity is required for such bulk operations, there are two potential solutions:

1.Design the code structure for bulk addition and deletion together, abstracting out the rollback logic.
2.Record both successful and failed parts of the operation and return them to the user. Ideally, provide the user with the SQL statements for failed partition operations as well."

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+                        if (stateMgr.getGlobalTransactionMgr()
+                                .existCommittedTxns(db.getId(), olapTable.getId(), partition.getId())) {
+                            throw new DdlException(
+                                    "There are still some transactions in the COMMITTED state waiting to be completed." +
```
It's a good error message.
And, it's recommended to put it into ErrorCode.java if it will be shown to users.
and, you can find proper SQLState in https://www.postgresql.org/docs/15/errcodes-appendix.html.
Maybe, the SQLState for this message could be 01000

jaogoy commented on fe/fe-core/src/main/java/com/starrocks/server/LocalMetastore.java:
```diff
@@ -1563,59 +1563,120 @@ public void replayAddPartition(PartitionPersistInfo info) throws DdlException {
+                ErrorReport.reportDdlException(ErrorCode.ERR_DROP_PARTITION_NON_EXISTENT, notExistPartition.toString());
             }
+        }
+        for (String partitionName : existPartition) {
```
Is the droping command atomic?

jaogoy reviewed (commented):
I think the syntax is OK.
and later, we can implement more similar clause such as `DORP PARTITIONS p1, p2, ...` and `DROP PARTITIONS (v11, v12), (v21, v22), ...`;

luohaha committed:
[UT] fix crash ut PersistentIndexSstableTest.test_merge (#43073)

Signed-off-by: luohaha <18810541851@163.com>

sduzh committed:
[BugFix] Fix incorrect partition id in log (#43079)

Signed-off-by: Alex Zhu <zhuming9011@gmail.com>

Smith-Cruise committed:
[BugFix] Fixed external catalog's PartitionKey being deduplicated (#42893)

Signed-off-by: Smith Cruise <chendingchao1@126.com>

sduzh committed:
[BugFix] Fix partition deletion keeps failing due to non-existent shard (#43083)

Signed-off-by: Alex Zhu <zhuming9011@gmail.com>

murphyatwork committed:
[BugFix] refactor the error message when mv is inactivated (#42948)

Signed-off-by: Murphy <mofei@starrocks.com>

nshangyiming committed:
[BugFix] Disable set password for non-native user (#42579)

Signed-off-by: Dejun Xia <xiadejun@starrocks.com>

trueeyu committed:
[BugFix] Function framework support checkout overflow (#43065)

## Why I'm doing:

Some functions, such as `concat`,  `bitmap_to_string`, will generate a `BinaryColumn` exceeding 4G. There is no check in the current function framework, which may lead to OOM or invalid result.

```
mysql> select sum(length(cast(c1 as string))) from t2;                                                                                                                                                                                                                            
+-----------------------------------------+                                                                                                                                                                                                                                       
| sum(length(CAST(c1 AS VARCHAR(65533)))) |                                                                                                                                                                                                                                       
+-----------------------------------------+                                                                                                                                                                                                                                       
|                              5520000000 |                                                                                                                                                                                                                                       
+-----------------------------------------+
```

Before the pr
```
mysql> select length(bitmap_to_string(bitmap_agg(c1))) from t2;                                                                                                                                                                                                                   
+------------------------------------------+                                                                                                                                                                                                                                      
| length(bitmap_to_string(bitmap_agg(c1))) |                                                                                                                                                                                                                                      
+------------------------------------------+                                                                                                                                                                                                                                      
|                               1705032703 |                                                                                                                                                                                                                                      
+------------------------------------------+ 
```

After the pr
```
mysql> select length(bitmap_to_string(bitmap_agg(c1))) from t2;                                                                                                                                                                                                                   
ERROR 1064 (HY000): Result column of function bitmap_to_string exceed limit: Total byte size of binary column exceed the limit: 4294967296 
```

## What I'm doing:

Determine whether to check overflown based on function configuration.

List of functions that currently require check overflow:

```
space
repeat
append_trailing_char_if_absent
concat
concat_ws
array_concat_ws
split_part
split
regexp_replace
replace_old
replace
to_binary
from_binary
bitmap_to_string
bitmap_to_base64
bitmap_to_binary
get_json_string
get_json_object
json_string
to_base64
array_join
array_concat
```

The corresponding test case will be added to StarRocksTest.

Signed-off-by: trueeyu <lxhhust350@qq.com>

leiyang0324 commented on fe/fe-core/src/main/java/com/starrocks/alter/AlterJobMgr.java:
```diff
@@ -575,7 +575,7 @@ public void processAlterTable(AlterTableStmt stmt) throws UserException {
                         DynamicPartitionUtil.checkAlterAllowed((OlapTable) db.getTable(tableName));
                     }
-                    if (dropPartitionClause.getPartitionName()
+                    if (dropPartitionClause.getPartitionName() != null && dropPartitionClause.getPartitionName()
```
> Should we take consideration of the `partitionNames`?

Thank you ，Is there a scenario for bulk deletion of shadow partitions?

andyziye committed:
[Tool] Request approval for feature (#43075)

Signed-off-by: AndyZiYe <yeziyu@starrocks.com>

lianneli committed:
[Doc] keep the same content with English Doc 'backup_restore' (#43068)

Signed-off-by: ming.li2 <ming.li2@dmall.com>

packy92 committed:
[Enhancement] support more character in db/col name (#42954)

Signed-off-by: packy92 <wangchao@starrocks.com>

andyziye committed:
[Tool] Fix the workflow not be triggered by GITHUB_TOKEN (#43128)

Signed-off-by: AndyZiYe <yeziyu@starrocks.com>

nshangyiming committed:
[BugFix] Fix compatability issue when ipv6 addr is used (#43078)

Signed-off-by: Dejun Xia <xiadejun@starrocks.com>

wangsimo0 committed:
[Doc]Update SHOW_AUTHENTICATION.md (#42957)

Signed-off-by: simo <48942089+wangsimo0@users.noreply.github.com>
Signed-off-by: evelyn.zhaojie <everlyn.zhaojie@gmail.com>
Co-authored-by: evelyn.zhaojie <everlyn.zhaojie@gmail.com>

wangsimo0 committed:
[Doc]Update SHOW_AUTHENTICATION.md (#42960)

Signed-off-by: simo <48942089+wangsimo0@users.noreply.github.com>
Signed-off-by: evelyn.zhaojie <everlyn.zhaojie@gmail.com>
Co-authored-by: evelyn.zhaojie <everlyn.zhaojie@gmail.com>

amber-create committed:
[Doc] update delta lake faq (#43143)

andyziye committed:
[Tool] Replace restful api with client to request PR reviewer (#43150)

Signed-off-by: AndyZiYe <yeziyu@starrocks.com>

trueeyu committed:
[BugFix] SyncPoint only takes effect in unit test (#43124)

Signed-off-by: trueeyu <lxhhust350@qq.com>

Seaven committed:
[BugFix] to compatible array_distinct id (#43146)

Signed-off-by: Seaven <seaven_7@qq.com>

gengjun-git committed:
[UT] Fix PseudoBackend report tablets bug (#43165)

Report_version must be set before setting the tablets to ensure that the tablet in FE will not be accidentally deleted in the following case:
t1: tablet1, tablet2, tablet3 are created on BE and current reportVersion is 3.
t2: tablet1, tablet2, tablet3 are set into request.tables.
t3: tablet4 is created and reportVersion changed to 4.
t4: 4 is set into request.report_version.
t5: The request is sent to the FE node and it is found that tablet4 is not in request.tablets
and request.report_version is equal to the latest report_version recorded by FE,
tablet4 metadata will be deleted from FE. Code: ReportHandler.deleteFromMeta

Signed-off-by: gengjun-git <gengjun@starrocks.com>

satanson committed:
[BugFix] Remove workgroup_id and version pair from workgroup_versions if the workgroup is removed (#43160)

Signed-off-by: satanson <ranpanf@gmail.com>

chaixiaoxue committed:
[Enhancement] Show create table and desc table add field comment info for external table (#37712)

Signed-off-by: chaixiaoxue <1156531142@qq.com>

kevincai committed:
[Enhancement] update staros version to v3.2-rc9 (#43180)

Signed-off-by: Kevin Xiaohua Cai <caixiaohua@starrocks.com>

wangsimo0 committed:
[Doc]  Update default value of `enable_recursive_listing` in hive_catalog and file external table (#42979)

Signed-off-by: simo <48942089+wangsimo0@users.noreply.github.com>
Signed-off-by: evelynzhaojie <everlyn.zhaojie@gmail.com
Co-authored-by: evelyn.zhaojie <98087056+evelynzhaojie@users.noreply.github.com>

wangsimo0 committed:
[Doc] Update default value of `enable_recursive_listing` in hive_catalog and file external table (#42977)

Signed-off-by: simo <48942089+wangsimo0@users.noreply.github.com>
Signed-off-by: evelynzhaojie <everlyn.zhaojie@gmail.com
Co-authored-by: evelyn.zhaojie <98087056+evelynzhaojie@users.noreply.github.com>

mxdzs0612 committed:
[Enhancement] Scale writers for connector sink when writing static partition tables (#40540)

Signed-off-by: Jiao Mingye <mxdzs0612@gmail.com>

LiShuMing committed:
[Enhancement] Support transparent union rewrite(Part 1) (#42541)

Signed-off-by: shuming.li <ming.moriarty@gmail.com>

zhangbutao committed:
[Enhancement] Skip to write Iceberg table/column comment if comment is empty (#43127)

Signed-off-by: zhangbutao <zhangbutao@cmss.chinamobile.com>

andyziye committed:
[Tool] Fix thirdparty update in merge pipeline (#43202)

Signed-off-by: AndyZiYe <yeziyu@starrocks.com>

fieldsfarmer committed:
[Tool] Update tool chain and dev env dockerfiles (#43100)

Signed-off-by: Kaiwei Wang <k.wang@celonis.com>

letian-jiang committed:
[BugFix] Fix parquet writer being closed when there is still unfinished task (#43121)

Signed-off-by: Letian Jiang <letian.jiang@outlook.com>

amber-create committed:
[Doc] fix path inconsistency in code in stream load user guide doc (#43210)

evelynzhaojie committed:
[Doc] fix links in flink-connector-starrocks (#43226)

Signed-off-by: evelynzhaojie <everlyn.zhaojie@gmail.com

sduzh committed:
[BugFix] Fix forgot to unlock database lock in LakeTableAlterMetaJobBase (#43155)

Signed-off-by: Alex Zhu <zhuming9011@gmail.com>

gengjun-git committed:
[Enhancement] Add retry to push image (#40939)

Signed-off-by: gengjun-git <gengjun@starrocks.com>

TszKitLo40 committed:
[BugFix] Fix bug in schema change with sort key in shared_data mode (#43204)

Signed-off-by: Zijie Lu <wslzj40@gmail.com>

leiyang0324 committed:
Modify the logic of replayDropPartition,add batch deletion of new syntax

leiyang0324 committed:
Modify the logic of replayDropPartition,add batch deletion of new syntax

MicePilot committed:
[Feature] Support typeof function (#42627)

The typeof function is completely implemented in fe.
In ExpressionAnalyzer.visitFunctionCall, we can get the parameter type of typeof. At this time, we have already got the result of the typeof function. We only need to directly replace the current FunctionCallExpr with StringLiteral, but I can't get the parent node of typeof, so I can't make this replacement. To solve this problem, I put StringLiteral in the position of the typeof parameter, and then rewrite it in the FoldConstantsRule.

Signed-off-by: changxin <streakxin@foxmail.com>

letian-jiang committed:
[Enhancement] Set thread local information for async task executed in IO threads pool (#43209)

Why I'm doing:
To ease debugging, set thread local information for async task executed in IO threads pool

What I'm doing:
Set query_id and fragment_instance_id

Signed-off-by: Letian Jiang <letian.jiang@outlook.com>

decster committed:
[UT] Remove BeRestartNewPublishTest (#43242)

Why I'm doing:
BeRestartNewPublishTest takes long time to run, and there is already a BeRestartTest with similar function, and new publish is not actively used.

What I'm doing:
Remove BeRestartNewPublishTest

Signed-off-by: Binglin Chang <decstery@gmail.com>

nshangyiming committed:
[UT] Make ut always bind to ipv4 loopback addr (#43253)

Signed-off-by: Dejun Xia <xiadejun@starrocks.com>

evelynzhaojie committed:
[Doc] optimize example in round(), curdate() (#43254)

Signed-off-by: evelynzhaojie <everlyn.zhaojie@gmail.com

LiShuMing committed:
[BugFix] Fix drop normal table failed when table has synchronized mv (#41990)

Signed-off-by: shuming.li <ming.moriarty@gmail.com>

dirtysalt committed:
[BugFix] fix excessive io task in adaptive strategy (#43183)

Signed-off-by: yanz <dirtysalt1987@gmail.com>

dirtysalt committed:
[BugFix] fix wrong impl of `consine_similarity` (#43274)

trueeyu committed:
[BugFix] Fix some problems of skew join (#43199)

Signed-off-by: trueeyu <lxhhust350@qq.com>

kevincai committed:
[BugFix] fix mem detect error in cgroup2fs env (#43270)

Signed-off-by: Kevin Xiaohua Cai <caixiaohua@starrocks.com>

leiyang0324 committed:
revert checkstyle-header.txt changes

leiyang0324 committed:
revert checkstyle-header.txt changes

zombee0 committed:
[Feature] impl left join unnest on true (#42924)

Signed-off-by: zombee0 <ewang2027@gmail.com>

mergify[bot] committed:
[Tool] Request approval for feature in branch (backport #43288) (#43290)

Signed-off-by: AndyZiYe <yeziyu@starrocks.com>

packy92 committed:
[BugFix] fix projection ref non-exist cols (#43262)

Signed-off-by: packy92 <wangchao@starrocks.com>

sduzh committed:
[UT] Fix flaky test RollupJobV2Test.testSchemaChange1 (#43297)

Signed-off-by: Alex Zhu <zhuming9011@gmail.com>

hellolilyliuyi committed:
[Doc]add more description in kafka connector and modify the name for op field (#43074)

Signed-off-by: hellolilyliuyi <96421222+hellolilyliuyi@users.noreply.github.com>

smartlxh committed:
[Enhancement] Support load segment parallel in shared_data (#36964)

Signed-off-by: smartlxh <smartlxh@gmail.com>
Signed-off-by: lixianhai <lixianhai.lxh@alibaba-inc.com>

hellolilyliuyi committed:
[Doc]change table format to list format in hybrid-table (#43308)

Signed-off-by: hellolilyliuyi <96421222+hellolilyliuyi@users.noreply.github.com>

luohaha committed:
[Enhancement] avoid useless compaction in cloud native pk table (#43271)

Signed-off-by: luohaha <18810541851@163.com>

HangyuanLiu committed:
[Refactor] Move some server and manager to GlobalStateMgr (#43093)

HangyuanLiu committed:
[BugFix] Unify and standardize the default implementation of AstVisitor (#43311)

Signed-off-by: HangyuanLiu <460660596@qq.com>

Youngwb committed:
[BugFix] Fix skew join unknown error when use skew join hint (#43319)

Signed-off-by: Youngwb <yangwenbo_mailbox@163.com>

before-Sunrise committed:
[BugFix] fix hdfs scanner cpu time negative problem (#43305)

Signed-off-by: before-Sunrise <unclejyj@gmail.com>

zombee0 committed:
[Refactor]multi_unnest deal with null row and refact unnest offset logic (#43296)

Signed-off-by: zombee0 <ewang2027@gmail.com>

Seaven committed:
[Enhancement] check compare optmization more strict (#42936)

Signed-off-by: Seaven <seaven_7@qq.com>

imay committed:
Update PULL_REQUEST_TEMPLATE.md (#43353)

Signed-off-by: imay <buaa.zhaoc@gmail.com>

HangyuanLiu committed:
[Feature] Support use warehouse id to get compute node (#42941)

Signed-off-by: HangyuanLiu <460660596@qq.com>

mxdzs0612 committed:
[BugFix] Fix wrong cast from double to decimal when using "create table as select" statement to create hive table (#42350)

Signed-off-by: Jiao Mingye <mxdzs0612@gmail.com>

xiangguangyxg committed:
[Enhancement] Improve replication job scheduling to support heterogeneous replica number and improve fault tolerance (#42772)

Signed-off-by: xiangguangyxg <xiangguangyxg@gmail.com>

srlch committed:
[BugFix] Fix incorrect filtering of complex expression predicates by GIN (#43153)

Signed-off-by: srlch <linzichao@starrocks.com>

dirtysalt committed:
[UT] fix sql test bad result (#43360)

Signed-off-by: yanz <dirtysalt1987@gmail.com>

dirtysalt committed:
[Refactor] make udf error message clearer (#43275)

Signed-off-by: yanz <dirtysalt1987@gmail.com>

imay committed:
Update PULL_REQUEST_TEMPLATE.md (#43364)

Signed-off-by: imay <buaa.zhaoc@gmail.com>

banmoy committed:
[Feature] Support LoadChannel profile (#41664)

Signed-off-by: PengFei Li <lpengfei2016@gmail.com>

starrocks-xupeng committed:
[Enhancement] adjust configs for cloud star cache (#43320)

Signed-off-by: starrocks-xupeng <xupeng@starrocks.com>

sevev committed:
[BugFix] Disable duplicate sort key when creating table (#43206)

Signed-off-by: zhangqiang <qiangzh95@gmail.com>

luohaha committed:
[UT] add test for LakePrimaryKeyCompactionTest.test_compaction_policy_min_input (#43351)

Signed-off-by: luohaha <18810541851@163.com>

letian-jiang committed:
[BugFix] Fix iceberg sink with null partition (#43321)

Signed-off-by: Letian Jiang <letian.jiang@outlook.com>

zhangbutao committed:
[Enhancement] Make IcebergHiveCatalog's list-all-tables property configurable (#43069)

Signed-off-by: zhangbutao <zhangbutao@cmss.chinamobile.com>

murphyatwork committed:
[Enhancement] add more trace for mv preprocess (#43361)

Signed-off-by: Murphy <mofei@starrocks.com>

sduzh committed:
[BugFix] Fix forgot to update max column unique id (#43352)

Signed-off-by: Alex Zhu <zhuming9011@gmail.com>

Smith-Cruise committed:
[Refactor] Remove duplicate add partition logic in HDFS text reader (#43310)

Signed-off-by: Smith Cruise <chendingchao1@126.com>

smartlxh committed:
[Feature] Support LakeTablet tablet_internal_parallel (#41348)

Signed-off-by: smartlxh <smartlxh@gmail.com>
Signed-off-by: lixianhai <lixianhai.lxh@alibaba-inc.com>

srlch committed:
[BugFix] Fix shutdown tablet meta leak cause migration fail. (#42827)

## Why I'm doing:
Suppose we have there different datadir called dir1 dir2 dir3. One of the problematic execution timeline is following:
1. Create a tablet with id 123 on dir1, we call it tablet_123_dir1.

2. migrate it into dir2, tablet_123_dir1 will be added into shutdown_tablets map but will not remove the meta data in rocksdb immediately. And we re-create a tablet_123_dir2 on dir2, migration is done.

3. migrate it into dir3, tablet_123_dir2 will be try to added into shutdown_tablets. But problem happen here: we actully can not add tablet_123_dir2 into shutdown_tablets because we use std::map::emplace to add the tablet, but <123, tablet_123_dir1> has already existed in this map, std::map::emplace will not rewrite <123, tablet_123_dir1> into <123, tablet_123_dir2>. So we actully destory tablet_123_dir2 immediately without remove its meta data. And the meta data for tablet_123_dir2 in rocksdb can not be removed forever in this case.

4. migrate it back to dir2. Now we notice that there is tablet meta data in dir2 for tablet 123. But we can not remove it this case, because tablet_123_dir2 is gone.

The core problem here is that shutdown_tablets is std::map and we use std::map::emplace to add a new shutdown tablet. It may leak if the tablet id has been existed in this map.

## What I'm doing:
Forcely remove the previous shutdown tablet with same tablet id if add a new one to it.


Signed-off-by: srlch <linzichao@starrocks.com>

wyb committed:
[BugFix] Fix random distribution type and dynamic partition properties in tables_config table (#43222)

Signed-off-by: wyb <wybb86@gmail.com>

smartlxh committed:
[Feature] Support io coalesce in shared_data (#39276)

Signed-off-by: smartlxh <smartlxh@gmail.com>
Signed-off-by: lixianhai <lixianhai.lxh@alibaba-inc.com>

Youngwb committed:
[Enhancement] Use statistics to optimize skew join (#43166)

Signed-off-by: Youngwb <yangwenbo_mailbox@163.com>

amber-create committed:
[Doc] remove Load data from cloud storage (#43380)

letian-jiang committed:
[BugFix] Handle file writer init error correctly (#43287)

Signed-off-by: Letian Jiang <letian.jiang@outlook.com>

HangyuanLiu committed:
[Refactor] Delete Create Security Integration Statement related code (#43354)

Signed-off-by: HangyuanLiu <460660596@qq.com>

nshangyiming committed:
[BugFix] Fix ipv6 + fqdn in dual-stack network environment (#43322)

Signed-off-by: Dejun Xia <xiadejun@starrocks.com>

nshangyiming commented:
@leiyang0324  could you rebase the main branch to resolve conflicts, so that we can run the ut.

before-Sunrise committed:
[UT] delete unstable test (#43392)

Signed-off-by: before-Sunrise <unclejyj@gmail.com>

amber-create committed:
[Doc] add 3.1.10 release notes (#43405)

Smith-Cruise committed:
[Doc] Update JAVA_UDF.md (#43402)

Signed-off-by: Smith Cruise <chendingchao1@126.com>
Signed-off-by: evelynzhaojie <everlyn.zhaojie@gmail.com
Co-authored-by: evelyn.zhaojie <98087056+evelynzhaojie@users.noreply.github.com>

gengjun-git committed:
[BugFix] Set max open files to 65535 when starting BE (#43404)

Signed-off-by: gengjun-git <gengjun@starrocks.com>

amber-create committed:
[Doc] Update 3.1.10 release notes (#43416)

Signed-off-by: amber-create <48005258@qq.com>

dirtysalt committed:
[BugFix] shorten jdbc driver name to avoid file name too long (#43277)

Signed-off-by: yanz <dirtysalt1987@gmail.com>

murphyatwork committed:
[BugFix] fix union mv when the partitions is unaligned (#42950)

Signed-off-by: Murphy <mofei@starrocks.com>

andyziye committed:
[Tool] Init branch-3.3 (#43425)

Signed-off-by: AndyZiYe <yeziyu@starrocks.com>

leiyang0324 commented:
> @leiyang0324 could you rebase the main branch to resolve conflicts, so that we can run the ut.

ok,i will resolve conflicts

leiyang0324 committed:
Merge remote-tracking branch 'origin/batch-drop-partitions' into batch-drop-partitions

# Conflicts:
#	fe/fe-core/src/main/java/com/starrocks/common/ErrorCode.java

nshangyiming commented:
![image](https://github.com/StarRocks/starrocks/assets/107105845/908ae3dc-480d-4cd7-a7ec-0c91ebb1e008)

@leiyang0324  there must be something wrong...

leiyang0324 commented:
> ![image](https://private-user-images.githubusercontent.com/107105845/318325776-908ae3dc-480d-4cd7-a7ec-0c91ebb1e008.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTE5NDE2ODQsIm5iZiI6MTcxMTk0MTM4NCwicGF0aCI6Ii8xMDcxMDU4NDUvMzE4MzI1Nzc2LTkwOGFlM2RjLTQ4MGQtNGNkNy1hN2VjLTBjOTFlYmIxZTAwOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQwMVQwMzE2MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOWMyMmJiMjVlNDBjMGI4OGZlMDk3M2M0NTc5N2M2ZDQ0OTZkZDMwM2QxZDZiM2NmZGExMjRjNjJkYWU1NzRmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.bJbTvVbzZE8PcJjgAlV54FHMdikMjcRhduJ180PyjXM)
> 
> @leiyang0324 there must be something wrong...
I will pull a new branch from the latest master branch, apply the code patch for batch partition deletion to this new branch, and then create a new pull request.

nshangyiming commented:
Going to close this pr due to rebase, switch to this new pr: https://github.com/StarRocks/starrocks/pull/43539.
